### PR TITLE
En/review eg/coil perturbations fix

### PR DESCRIPTION
**Pull request by eduardolneto**
Changing coil_perturbation and coils for coil perturbation and stochastic optimization functionalities to comply with custom_losses

### CODE REVIEW

File: essos/coil_perturbation.py [hunk 1]
```diff
@@ -205,70 +205,92 @@ def get_sample(self, deriv):
...
-        curves.gamma_dash=curves.gamma_dash + gamma_perturbations_dash        
-        curves.gamma_dashdash=curves.gamma_dashdash + gamma_perturbations_dashdash
+        base_gamma = Curves(curves.dofs, curves.n_segments, nfp=1, stellsym=False).gamma
+        perturbed_base_gamma = base_gamma + perturbation[:, 0, :, :]
+        dofs_new, _ = fit_dofs_from_coils(perturbed_base_gamma, curves.order, curves.n_segments,assume_uniform=True)
+        curves.dofs = dofs_new
+        return
+
+    raise TypeError(f"Unsupported type {type(curves)}. Expected Curves, Coils, or CoilsFromGamma.")
     #return curves
```
Comment (EstevaoMGomes) [inline essos/coil_perturbation.py]: Remove

File: essos/coil_perturbation.py [hunk 2]
```diff
@@ -205,70 +205,92 @@ def get_sample(self, deriv):
...
-        gamma_perturbations_dash = apply_symmetries_to_gammas(perturbation[:,1,:,:], curves.nfp, curves.stellsym)
-        gamma_perturbations_dashdash = apply_symmetries_to_gammas(perturbation[:,2,:,:], curves.nfp, curves.stellsym)        
-        curves.gamma=curves.gamma + gamma_perturbations    
-        curves.gamma_dash=curves.gamma_dash + gamma_perturbations_dash        
-        curves.gamma_dashdash=curves.gamma_dashdash + gamma_perturbations_dashdash
+        base_gamma = Curves(curves.dofs, curves.n_segments, nfp=1, stellsym=False).gamma
+        perturbed_base_gamma = base_gamma + perturbation[:, 0, :, :]
+        dofs_new, _ = fit_dofs_from_coils(perturbed_base_gamma, curves.order, curves.n_segments,assume_uniform=True)
+        curves.dofs = dofs_new
+        return
```
Comment (EstevaoMGomes) [inline essos/coil_perturbation.py]: I thought the source code was implemented in a simpler way, like:

```
dofs, skeleton = ravel_pytree(curves)
new_dofs = dofs + perturbation
new_curves = skeleton(new_dofs)
```
where perturbation is a vector with the same shape as dofs.

Comment (EstevaoMGomes) [reply essos/coil_perturbation.py]: Btw, this creates a new object instead of altering the last one in-place. I think this is better.

File: essos/coil_perturbation.py [hunk 3]
```diff
@@ -205,70 +205,92 @@ def get_sample(self, deriv):
...
+
+    raise TypeError(f"Unsupported type {type(curves)}. Expected Curves, Coils, or CoilsFromGamma.")
     #return curves  
 
 
-def perturb_curves_statistic(curves: Curves,sampler:GaussianSampler, key=None):
+def perturb_curves_statistic(curves, sampler:GaussianSampler, key=None):
```
Comment (EstevaoMGomes) [inline essos/coil_perturbation.py]: This works, but since both functions have similar implementations, wouldn't it make more sense to add another input to indicate whether the perturbation is statistical or systematic?

File: essos/coils.py [hunk 4]
```diff
@@ -25,7 +25,15 @@ def __init__(self,
                  dofs:        jnp.ndarray,
                  n_segments:  int = 100,
                  nfp:         int = 1,
-                 stellsym:    bool = True):
+                 stellsym:    bool = True,
+                 scaling_type: int = 2,
+                 scaling_factor: float = 0,
```
Comment (EstevaoMGomes) [inline essos/coils.py]: Set 0.0 as it is a float

File: essos/coils.py [hunk 5]
```diff
@@ -41,6 +49,11 @@ def __init__(self,
         self._nfp = nfp
         self._stellsym = stellsym
 
+        self._scaling_type = scaling_type  # 1 for L-1 norm, 2 for L-2 norm, jnp.inf for L-infinity norm
```
Comment (EstevaoMGomes) [inline essos/coils.py]: Maybe should be -1 for L-infinity

File: essos/coils.py [hunk 6]
```diff
@@ -61,12 +74,13 @@ def reset_cache(self):
     # dofs property and setter
     @property
     def dofs(self):
-        return jnp.array(self._dofs)
+        # Apply scaling to each coordinate (X, Y, Z) independently
```
Comment (EstevaoMGomes) [inline essos/coils.py:115]: This makes no sense. When using the new loss class, the dofs that are used are those that form the PyTree, i.e., `self._dofs`. When you scale only the dofs property, this is only useful for the old optimizations. I think the scaling is a good addition, but as it is implemented, it should not be merged.
Furthermore, it seems overcomplicated for the user.

File: essos/coils.py [hunk 7]
```diff
@@ -135,16 +199,8 @@ def create_data(order: int) -> jnp.ndarray:
     # gamma property
     @property
     def gamma(self):
-        # Allow downstream code (e.g. coil_perturbation) to override the
-        # computed gamma; otherwise compute from Fourier coefficients.
-        if self._gamma is not None:
-            return self._gamma
```
Comment (EstevaoMGomes) [inline essos/coils.py:141]: What is the reason for removing the delayed initialization? I am becoming more and more confused with these changes...

File: essos/coils.py [hunk 8]
```diff
@@ -690,7 +799,289 @@ def CreateEquallySpacedCurves(n_curves:   int,
     curves = curves.at[:, 1, 0].set(jnp.sin(angles) * R)  # y[0]
     curves = curves.at[:, 1, 2].set(jnp.sin(angles) * r)  # y[2]
     curves = curves.at[:, 2, 1].set(-r)                   # z[1] (constant for all)
-    return Curves(curves, n_segments=n_segments, nfp=nfp, stellsym=stellsym)
+    return Curves(curves, n_segments=n_segments, nfp=nfp, stellsym=stellsym, scaling_type=scaling_type, scaling_factor=scaling_factor, scale_fixed=scale_fixed)
+
+def extract_axis_from_surface(surface, n_samples: int = 200):
```
Comment (EstevaoMGomes) [inline essos/coils.py]: Shouldn't this be in `surface.py` as a method of the class?

File: essos/coils.py [hunk 9]
```diff
@@ -690,7 +799,289 @@ def CreateEquallySpacedCurves(n_curves:   int,
...
+    
+    # Convert to Cartesian coordinates
+    X_axis = R_axis * jnp.cos(phi)
+    Y_axis = R_axis * jnp.sin(phi)
+    
+    axis_gamma = jnp.stack([X_axis, Y_axis, Z_axis], axis=1)  # (n_samples, 3)
+    
+    return axis_gamma
+
+def CreateCoilsAroundAxis(n_coils: int,
```
Comment (EstevaoMGomes) [inline essos/coils.py]: I am not saying this is wrong but the code looks way too complex for what it does...

File: essos/coils.py [hunk 10]
```diff
@@ -834,4 +1225,547 @@ def fit_dofs_from_coils(
         gamma_uni = _resample_closed_curve_uniform_batch(coils_gamma, n_segments)  # arclength (vmapped)
 
     dofs = _fit_real_fourier_batch(gamma_uni, order)  # rFFT-based fit
-    return dofs, gamma_uni
\ No newline at end of file
+    return dofs, gamma_uni
+
+class CoilsFromGamma:
```
Comment (EstevaoMGomes) [inline essos/coils.py]: Just so I know, is this supposed to be an optimizable or not?

Comment (EstevaoMGomes) [reply essos/coils.py]: I see it is registered as a pytree, so I suppose so. Then what are the dofs? Is it all the gammas?

**Review by EstevaoMGomes — COMMENTED** [review]
This PR does not seem master-ready. More exactly, the scaling implementation is an important feat, but it needs to be implemented properly.

File: essos/coils.py (hunk 6 above)
Comment (eduardolneto) [reply essos/coils.py:115]: I am not sure I understand the problem here.  Both the property and setter are scaled, as it is also done on surfaces.py, which was already implemented.

File: essos/coils.py (hunk 5 above)
Comment (eduardolneto) [reply essos/coils.py]: We can do that but I woudl think this is more confusing for the user? Surfaces.py was implemented like this as well, which I think is because the options used in jnp.linalg.norm functions are also these ones and not -1.

File: essos/coil_perturbation.py [hunk 11]
```diff
@@ -205,70 +205,92 @@ def get_sample(self, deriv):
 
 
 
-def perturb_curves_systematic(curves: Curves,sampler:GaussianSampler, key=None):
+def perturb_curves_systematic(curves, sampler:GaussianSampler, key=None):
     """
     Apply a systematic perturbation to all the coils. 
-    This means taht an independent perturbation is applied to the each unique coil
-    Then, the required symmetries are applied to the perturbed unique set of coils
```
Comment (rogeriojorge) [inline essos/coil_perturbation.py:212]: I actually liked these descriptions better, "systematic" is not well defined and this defined it pretty well

File: essos/coils.py [hunk 12]
```diff
@@ -177,14 +227,8 @@ def create_data(order: int) -> jnp.ndarray:
...
     @property
     def gamma_dashdash(self):
-        if self._gamma_dashdash is not None:
-            return self._gamma_dashdash
         return self._compute_gamma_dashdash()
 
-    @gamma_dashdash.setter
-    def gamma_dashdash(self, value):
-        self._gamma_dashdash = value
-
```
Comment (rogeriojorge) [inline essos/coils.py:187]: This looks important

File: essos/coils.py [hunk 13]
```diff
@@ -343,7 +387,7 @@ def wrap(data):
         polyLinesToVTK(str(filename), np.array(x), np.array(y), np.array(z), pointsPerLine=np.array(ppl), pointData=pointData)
 
     @classmethod
-    def from_simsopt(cls, simsopt_curves, nfp=1, stellsym=True):
+    def from_simsopt(cls, simsopt_curves, nfp=1, stellsym=True, scaling_type=2, scaling_factor=0.0):
```
Comment (rogeriojorge) [inline essos/coils.py]: Not sure what scaling type is? Perhaps should be given in the docstring

File: essos/coils.py (hunk 9 above)
Comment (rogeriojorge) [inline essos/coils.py]: This seems like something that would help with pyQSC, when optimizing for coils with the near-axis approach. But I would probably move this for a different PR, unless it is really needed for the coils perturbation PR.

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
A couple things I suggest:

- Remove unused variables such as old_scaling in the order setter and R_axis in the surface Frenet-frame helper.

- Curves use self._dofs, but Coils.to_simsopt() iterates over self.dofs_curves, which goes through the scaled Curves.dofs property. That can make exported SIMSOPT coils differ from the actual geometry when scaling is active.

- Curves.dofs now returns scaled DOFs, but _tree_flatten() still exposes self._dofs, so custom_losses/JAX tree operations will see the unscaled variables, not the scaled public dofs. Also, scale_fixed is not included in the PyTree aux data, and scaling_type is stored but not actually used when building self.scaling.

File: essos/coils.py (hunk 5 above)
Comment (EstevaoMGomes) [reply essos/coils.py]: If it is well documented, it is not confusing. Using jnp.inf may lead to problems as not all infinities are the same (e.g., np.inf vs jnp.inf)

File: essos/coils.py (hunk 6 above)
Comment (EstevaoMGomes) [reply essos/coils.py:115]: The property is different from the attribute. `_tree_flatten()` gets the property, not the attribute. The scaling is just wrong as is.

Comment (eduardolneto) [reply essos/coils.py:115]: I understand that. But this is exactly what previous implemented scalling of surfaces.py do as well, which is what this is mimicking. @EstevaoMGomes

Comment (eduardolneto) [reply essos/coils.py:115]: Since that behaviour for surfaces comes from the original branch. Do you want to change both here or chnag eit after this has been pulled to the original branch?

File: essos/coil_perturbation.py (hunk 2 above)
Comment (eduardolneto) [reply essos/coil_perturbation.py]: @EstevaoMGomes I am not sure what you want with this comment. the purturbation is usually in cosntruction of the coils, so it is usually given in terms of the gamma quantity and not in fourier coeficients. if the instance is curves_from_Gammas it does what you are suggesting.

File: essos/coils.py (hunk 4 above)
Comment (eduardolneto) [reply essos/coils.py]: Done

File: essos/coil_perturbation.py (hunk 3 above)
Comment (eduardolneto) [reply essos/coil_perturbation.py]: Done

File: essos/coil_perturbation.py (hunk 11 above)
Comment (eduardolneto) [reply essos/coil_perturbation.py:212]: Done

**Comment by eduardolneto** [comment]
@rogeriojorge @EstevaoMGomes I fixed the scaling dofs behaviour. It is explained in the commits summary. But basically, now the dofs are unscaled consistently during unflatten, and the scaled ones are flatten. This was changed for both surfaces and coils. This means that the loss is calculated with unscaled and the gradients against the scaled ones. For the coils this meant that the current_scale had to be given as an input or, if not, it is calculated at the beginning and frozen. Before it was updated every time, but this would mean that the gradient tracer would see d loss(scaling(sofs) d dofs) which woudl give wrong gradient. DiscretizedCoils now also do the same buit the geometry dofs are the unique gammas which are scaled always to the maximum value of X,Y,Z like currents.

**Comment by eduardolneto** [comment]
I think it should be good to merge. And then test more with the examples once merged into #29 .

File: essos/coils.py (hunk 13 above)
Comment (eduardolneto) [reply essos/coils.py]: Fixed.

File: essos/coils.py (hunk 9 above)
Comment (eduardolneto) [reply essos/coils.py]: Fixed.

File: essos/coils.py (hunk 10 above)
Comment (eduardolneto) [reply essos/coils.py]: Fixed.

File: essos/coils.py (hunk 9 above)
Comment (eduardolneto) [reply essos/coils.py]: Fixed.

File: essos/coils.py (hunk 8 above)
Comment (eduardolneto) [reply essos/coils.py]: Fixed.

File: essos/coils.py (hunk 6 above)
Comment (eduardolneto) [reply essos/coils.py:115]: Fixed.

File: essos/coils.py (hunk 5 above)
Comment (eduardolneto) [reply essos/coils.py]: Fixed.

File: essos/coil_perturbation.py (hunk 2 above)
Comment (eduardolneto) [reply essos/coil_perturbation.py]: Done.

File: essos/coil_perturbation.py (hunk 1 above)
Comment (eduardolneto) [reply essos/coil_perturbation.py]: Done

File: essos/coils.py (hunk 12 above)
Comment (eduardolneto) [reply essos/coils.py:187]: Addresed

File: essos/coils.py (hunk 7 above)
Comment (eduardolneto) [reply essos/coils.py:141]: Addresed, only discretized coils are eager because they are dofs

**Review by rogeriojorge — APPROVED** [review]
Looks good to me
